Issue #62: improve syntax for locking workspaces

yt-ms opened the pull request:
* [X] fix #(issue number)
* [X] description of feature/fix
* [X] tests added/passed
* [X] add an entry to the [next release](../CHANGELOG.rst)

Added `lock()` method to `ClientWorkspace` to use for `with` blocks in favour of using the client directly.  For the moment, the old way is still there but we should remove in a future release.

On free plans, workspace locking is not supported, so ignore lock failures (credits to [@maximveksler](https://github.com/maximveksler)).

-------

THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE Apache License v.2.0. YOU MAY OBTAIN A COPY OF THE LICENSE AT https://www.apache.org/licenses/LICENSE-2.0.

THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

yt-ms committed:
feat(API): lock workspace in `with` block through `lock()` method

yt-ms committed:
fix(build): safety error reported for pip < 21.1

Midnighter commented on setup.cfg:
```diff
@@ -42,6 +42,7 @@ install_requires =
     httpx ~= 0.16
     importlib_metadata; python_version <'3.8'
     ordered-set
+    pip >= 21.1
```
pip should not be a direct dependency. If this is about safety complaining, we need to rather set this inside tox' dependencies.

yt-ms replied:
Ah, yes - this didn't feel right when I did it.  Will look to do that instead.

Midnighter commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -99,7 +100,11 @@ def __repr__(self) -> str:
-        """Enter a context by locking the corresponding remote workspace."""
+        """Enter a context by locking the corresponding remote workspace.
+
+        Note: this method is deprecated in favour of lock(), and will be removed in a
+        future relesae.
```
Sphinx has [a nice directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated) for this sort of thing. I haven't found anything like it for MkDocs yet. Either way, I suspect that since it's a dunder method, hardly anyone will look at the docstring. Instead, I suggest to create a `DeprecationWarning` when this method is called.

```py
warnings.warn("Using the `StructurizrClient` in a context is deprecated since version <next release> and will be removed in a future release. Please use the preferred method `.lock()` instead.", DeprecationWarning)
```

yt-ms replied:
0.4.1 for next release?

Midnighter replied:
Let's make it 0.5

Midnighter commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -109,14 +114,36 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
-        """Exit a context by unlocking the corresponding remote workspace."""
+        """Exit a context by unlocking the corresponding remote workspace.
+
+        Note: this method is deprecated in favour of lock(), and will be removed in a
+        future relesae.
```
If we create a deprecation warning to the user in `__enter__`, we can just delete this note.

Midnighter commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -109,14 +114,36 @@ def __enter__(self):
             )
 
+    @contextmanager
+    def lock(self):
```
Can't decide whether I'd like this name to be `lock` or `lock_context`... 🤔 I suppose `lock` is consistent with other class interfaces.

Midnighter replied:
or possibly `auto_lock`?

yt-ms replied:
Not sure I like `auto_lock`, as it doesn't make it automatic!  We could go for `workspace_lock()` as that reads quite well, but that might get confusing versus the `lock_workspace()` method.  If we do go for `workspace_lock()` then I'd be tempted to make that a property rather than a method:

    with client.workspace_lock:
        # Do some stuff

vs

    with client.lock():
        # Do some stuff

Midnighter replied:
Let's keep it as is then 🙂

Midnighter commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -109,14 +114,36 @@ def __enter__(self):
+    @contextmanager
+    def lock(self):
+        """Provide a context manager for locking and unlocking a workspace."""
+        is_successful = self.lock_workspace()
+        if not is_successful:
+            raise StructurizrClientException(
+                f"Failed to lock the Structurizr workspace {self.workspace_id}."
+            )
```
I would like to integrate logic from #73 here and only log a warning if locking is not successful due to being on the free plan.

yt-ms replied:
Yep, that was my intention once I got the basics working.

Midnighter commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -109,14 +114,36 @@ def __enter__(self):
+                f"Failed to lock the Structurizr workspace {self.workspace_id}."
+            )
+        try:
+            yield None
```
I think for consistency we should yield `self` here even though this value will most often be ignored.

```suggestion
            yield self
```

Midnighter commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -109,14 +114,36 @@ def __enter__(self):
+        finally:
+            is_successful = self.unlock_workspace()
+            self._client.close()
+            if not is_successful:
+                raise StructurizrClientException(
+                    f"Failed to unlock the Structurizr workspace {self.workspace_id}."
+                )
```
Same story as above, being on the free plan should not be an error.

Midnighter reviewed (changes requested):
Hey, thanks for taking this on! 🙂  Couple of wishes for changes in behaviour. I can also commit those changes myself if you prefer.

In addition to my comments, let's change `examples/upload_workspace.py` to not use a context.

yt-ms committed:
test: check exception handling for locks

yt-ms committed:
fix: move pip version constraint to tox.ini

yt-ms committed:
fix: warn properly on calling deprecated methods

yt-ms committed:
fix(api): ignore workspace locking errors on free plans

Creds to @maximveksler

yt-ms committed:
docs: update changelog

yt-ms committed:
refactor: update upload_workspace example to use lock()

yt-ms commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -192,25 +228,41 @@ def put_workspace(self, workspace: Workspace) -> None:
         response = self._client.send(request)
-        logger.debug("%r", response.json())
         response.raise_for_status()
+        logger.debug("%r", response.json())
```
FYI, the reason I swapped these lines over is that if the response code isn't 200 (e.g. because you gave a bad URL) then you don't get JSON in the response, but the real error gets masked by an exception trying to parse the JSON.

yt-ms replied:
I added a unit test that demonstrates that.

Midnighter replied:
Great 👍🏼

yt-ms commented:
It'd be very good if someone could try the free plan locking for real against Structurizr, as my account is bound to a paid plan rather than free.

Midnighter commented:
I can do that. Still have an e-mail address left 😛

yt-ms committed:
fix: bump "next version" to 0.5 in warnings

Midnighter commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -192,25 +228,41 @@ def put_workspace(self, workspace: Workspace) -> None:
+        """
+        success, paid_plan = self._lock_workspace()
+        if not success and not paid_plan and self.ignore_free_plan_locking_errors:
+            logger.debug("Ignoring lock failure on free plan")
```
I wonder if this is better logged as a warning saying that you cannot lock on the free plan and remove the `ignore_free_plan_locking_errors` attribute? I don't see anyone ever setting that attribute to `False`. (I might be wrong, of course.)

yt-ms replied:
Yeah, I was in two minds about that.  I'll remove the attribute so it doesn't confuse.

Midnighter commented on src/structurizr/api/structurizr_client.py:
```diff
@@ -227,11 +279,15 @@ def unlock_workspace(self) -> bool:
             )
-        return response.success
+            if self.ignore_free_plan_locking_errors and not self._paid_plan(response):
+                logger.debug("Ignoring unlock failure on free plan")
```
Same as before, better to have a warning?

Midnighter reviewed (commented):
Looks very good already. I feel a bit hesitant about adding a new attribute since, to me, it doesn't look very useful.

yt-ms committed:
refactor: always ignore lock failures on free plans

yt-ms commented:
OK, I've removed the attribute now, and rebased back from main, so hopefully good to go.  GitHub won't let me rebase and merge as it thinks there's still a change requested but won't show me where!
